### PR TITLE
feat: Handle Unverified Users Trying to Log In

### PR DESCRIPTION
**Pull request by carissa-tang**
## Notion ticket link

[Handle Unverified Users Trying to Log In](https://www.notion.so/uwblueprintexecs/Handle-Unverified-Users-Trying-to-Log-In-223c5520a124475d893b5a8607a26ff7?pvs=4)

## Implementation description
* When unverified teacher tries to login, they are redirected to a new "Verify Your Email" screen
* Note: admins cannot login without being verified - they are auto verified when they set their password

**Changes**
* Return `emailVerified` field in `login` resolver
   * True if email is verified
* Only set cookies and localStorage values if user is verified  
* Add `sendEmailVerificationLink` resolver / mutation
* `UnverifiedUsers` component -> shown if unverified user tries to login

## Steps to test
1. Create a new teacher account - don't verify it
2. Try to log in

Note: if you enter the wrong password, the existing window alert pops up - should we handle this differently, as might be a security hazard?

https://github.com/uwblueprint/jump-math/assets/69697180/41e401eb-c18d-4328-9845-6a5b45b1ff3f

## What should reviewers focus on?
* UI matches Figma
* No breaking changes

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/auth/Login.tsx [hunk 1]
```diff
@@ -9,12 +9,13 @@ import { ADMIN_SIGNUP_IMAGE, TEACHER_SIGNUP_IMAGE } from "../../assets/images";
 import * as Routes from "../../constants/Routes";
 import { HOME_PAGE } from "../../constants/Routes";
 import AuthContext from "../../contexts/AuthContext";
-import { AuthenticatedUser } from "../../types/AuthTypes";
+import { AuthenticatedUser, VerifiableUser } from "../../types/AuthTypes";
```
Comment (carissa-tang) [inline frontend/src/components/auth/Login.tsx:12]:
```suggestion
import { VerifiableUser } from "../../types/AuthTypes";
```

File: frontend/src/components/auth/reset-password/ForgotPassword.tsx [hunk 2]
```diff
@@ -1,22 +1,16 @@
...
 
 import { RESET_PASSWORD } from "../../../APIClients/mutations/AuthMutations";
 import { GET_USERS_BY_ROLE } from "../../../APIClients/queries/UserQueries";
 import { UserResponse } from "../../../APIClients/types/UserClientTypes";
-import { ArrowBackOutlineIcon } from "../../../assets/icons";
 import {
   ADMIN_SIGNUP_IMAGE,
   TEACHER_SIGNUP_IMAGE,
 } from "../../../assets/images";
+import LoadingState from "../../common/LoadingState";
```
Comment (carissa-tang) [inline frontend/src/components/auth/reset-password/ForgotPassword.tsx:12]:
```suggestion
```

File: frontend/src/components/auth/UnverifiedUsers.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  const [sendEmailVerificationLink] = useMutation<{ resetPassword: boolean }>(
+    SEND_EMAIL_VERIFICATION_LINK,
+  );
+
+  return (
+    <AuthWrapper
+      form={
+        <ResendEmail
+          resendFunction={async () => {
+            await sendEmailVerificationLink({ variables: { email } });
```
Comment (jfdoming) [inline frontend/src/components/auth/UnverifiedUsers.tsx:20]: nit: probably don't need async-await here, can just return the promise directly

File: frontend/src/components/common/ResendEmail.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  return (
+    <>
+      <Text textAlign="center" textStyle="subtitle2">
+        Didn’t receive the email?{" "}
+        <Button
+          color="blue.300"
+          display="contents"
+          onClick={onResendClick}
+          style={{ font: "inherit" }}
```
Comment (jfdoming) [inline frontend/src/components/common/ResendEmail.tsx]: just a thought, maybe we could consider disabling the button/putting it in a loading state (or just making the click do nothing) while the mutation is running? just so people don't accidentally send multiple emails to themselves

**Review by jfdoming — APPROVED** [review]
Out of curiosity, why was the extra `VerifiableUser` type needed? LGTM though

File: frontend/src/components/auth/UnverifiedUsers.tsx (hunk 3 above)
Comment (carissa-tang) [reply frontend/src/components/auth/UnverifiedUsers.tsx:20]: there's a type-mismatch btwn `Promise<FetchResult<{ resetPassword: boolean; }>>` and `() => Promise<void>`. i figured keeping a more generic type and using the wrapper would be cleaner

File: frontend/src/components/common/ResendEmail.tsx (hunk 4 above)
Comment (carissa-tang) [reply frontend/src/components/common/ResendEmail.tsx]: good idea, thanks!

**Comment by carissa-tang** [comment]
> VerifiableUser

I didn't directly add the field to `AuthenticatedUser` or `AuthenticatedAdminOrTeacher` if that's your question bc it gets complicated with `AuthenticatedUser` being in local storage but we can't store booleans (also there's no need to store it). i think it also got messy with the union being nullable

**Pull request merged**